_posts/2023-02-03_medical-imaging-with-nix/medical-imaging-with-nix.Rmd

```diff
@@ -0,0 +1,233 @@
+my experience hasn't been so pleasant, with most issues I've encountered feeling almost ungoogleable, different learning styles I guess?
+
+The other place I've found nixpkgs to be frustrating to use for python, is upgrading. Unless you're maintaining your own branch of nixpkgs
+you are somewhat at the whims of other nixpkg contributors as to what gets upgraded when. And due to a potentially poor choice on strategies
+the most up-to-date version of nixpkgs may contain a large set of temporarily broken python packages. Often upgrading one package will 
+break many packages that depend on it, not only because the code becomes incompatible, but because it is fashionable in python packaging 
+to set strict version upper bounds, so the package won't even build (so we can't check if all the tests still pass with the new version). 
+So I find myself in the position of checking out old versions of files from nixpkgs to build my package overlays when I need to 
+downdate a certain dependency. This is tedious and I should probably switch to maintaining my own version of nixpkgs with my 
+downdates and modifications, but this makes it harder to share with others.
+
+So this points to the biggest advantage of conda over nix, when conda does not have a package it gets it from pypi, nixpkgs cannot fail over
+to grabbing from pypi and installing with pip, it also can't do dependency solving, if the nixpkgs version of a python lib isn't compatible with
+another, you have to go find a satisfactory version yourself (or cheat and [lie about the version requirements](https://github.com/LKS-CHART/medical-imaging-nix/blob/2e811b3eb9b931f708d5f4b3658506b496ce24a0/container/overlay.nix#L71)^
+[I was reminded after the first draft of this post about the existence of `pythonRelaxDeps`, a way to automatically soften version requirements,
+I may switch to use this in the future.]). This is painful. There are two nix projects that have aimed to address this
+problem, [mach-nix](https://github.com/DavHau/mach-nix) which has been abandoned^[although hopefully returning as part of https://github.com/nix-community/dream2nix in the future], and [poetry2nix](https://github.com/nix-community/poetry2nix) which may solve
+some of my woes but I haven't tried it yet.
+
+Another advantage alluded to already is speed. Compiling things takes time, so unless you're always getting prebuilt binaries from the
+nix servers you can be in for long build times. Nix is good at not duplicating build work, but sometimes rebuilding is unavoidable. Say
+you want to use a newer version of cuda, or gcc, most of your environment will need to be rebuilt^[ a colleague suggests that we create a build
+farm where a background process builds new versions of things so that there are prebuilt versions always available, but that is an infrastructure
+investment that hasn't felt worth it yet. Apparently the issue is that CUDA is technically not free software and by policy the nix
```

Comment:
You get to this CUDA issue but maybe it's better to mention above where I complained about it since otherwise the reader might be surprised at how slow Nix is, idk